File: pkg/usage/usage.go
// Package usage extracts token accounting and stream shape from OpenAI- and
// Anthropic-style response bodies (buffered or SSE).
package usage

import (
	"bytes"
	"encoding/json"
	"fmt"
)

type Accumulator struct {
	Prompt, Cached, Completion *float64
	FinishReason               string
	ContentChars               int
	ReasoningChars             int
	ToolCallDeltas             int
	Generated                  bool
}

func (a *Accumulator) FeedJSON(data []byte) {
	var object map[string]any
	if json.Unmarshal(data, &object) == nil {
		a.FeedObject(object)
	}
}

func (a *Accumulator) FeedSSELine(line []byte) {
	line = bytes.TrimSpace(line)
	if !bytes.HasPrefix(line, []byte("data:")) {
		return
	}
	payload := bytes.TrimSpace(line[len("data:"):])
	if len(payload) == 0 || bytes.Equal(payload, []byte("[DONE]")) {
		return
	}
	a.FeedJSON(payload)
}

func (a *Accumulator) FeedObject(object map[string]any) {
	if choices, ok := object["choices"].([]any); ok && len(choices) > 0 {
		if first, ok := choices[0].(map[string]any); ok {
			if reason, ok := first["finish_reason"].(string); ok {
				a.FinishReason = reason
			}
			for _, rawChoice := range choices {
				choice, ok := rawChoice.(map[string]any)
				if !ok {
					continue
				}
				node, _ := choice["delta"].(map[string]any)
				if len(node) == 0 {
					node, _ = choice["message"].(map[string]any)
				}
				if content, ok := node["content"].(string); ok {
					a.ContentChars += len([]rune(content))
					a.Generated = a.Generated || content != ""
				}
				if reasoning, ok := node["reasoning_content"].(string); ok {
					a.ReasoningChars += len([]rune(reasoning))
					a.Generated = a.Generated || reasoning != ""
				}
				if reasoning, ok := node["reasoning"].(string); ok {
					a.ReasoningChars += len([]rune(reasoning))
					a.Generated = a.Generated || reasoning != ""
				}
				if truthy(node["tool_calls"]) {
					a.ToolCallDeltas++
					a.Generated = true
				}
			}
		}
	}
	if reason, exists := object["stop_reason"]; exists && reason != nil {
		a.FinishReason = fmt.Sprint(reason)
	}
	if status, ok := object["status"].(string); ok && (status == "completed" || status == "incomplete" || status == "cancelled") {
		a.FinishReason = status
	}
	// Anthropic streams generated text/reasoning in a top-level delta object;
	// the Responses API uses a top-level delta string for output-text events.
	if delta, ok := object["delta"].(map[string]any); ok {
		for _, key := range []string{"text", "thinking", "reasoning", "partial_json"} {
			if value, ok := delta[key].(string); ok && value != "" {
				a.Generated = true
			}
		}
	} else if delta, ok := object["delta"].(string); ok && delta != "" {
		a.Generated = true
	}

	usage, _ := object["usage"].(map[string]any)
	if message, ok := object["message"].(map[string]any); ok {
		if messageUsage, ok := message["usage"].(map[string]any); ok && len(messageUsage) > 0 {
			usage = messageUsage
		}
	}
	if usage == nil {
		return
	}
	if value, ok := numberValue(usage["prompt_tokens"]); ok {
		a.Prompt = &value
		cachedFound := false
		if details, ok := usage["prompt_tokens_details"].(map[string]any); ok {
			if cached, ok := numberValue(details["cached_tokens"]); ok {
				a.Cached = &cached
				cachedFound = true
			}
		}
		if !cachedFound {
			if cached, ok := numberValue(usage["cached_tokens"]); ok {
				a.Cached = &cached
			}
		}
	}
	if value, ok := numberValue(usage["completion_tokens"]); ok {
		a.Completion = &value
	}
	if value, ok := numberValue(usage["input_tokens"]); ok {
		a.Prompt = &value
	}
	if value, ok := numberValue(usage["cache_read_input_tokens"]); ok {
		a.Cached = &value
	}
	if value, ok := numberValue(usage["output_tokens"]); ok {
		a.Completion = &value
	}
}

func FeedSSEChunk(accumulator *Accumulator, tail, chunk []byte) []byte {
	tail = append(tail, chunk...)
	for {
		idx := bytes.IndexByte(tail, '\n')
		if idx < 0 {
			return tail
		}
		accumulator.FeedSSELine(tail[:idx])
		tail = tail[idx+1:]
	}
}

func numberValue(value any) (float64, bool) {
	switch value := value.(type) {
	case float64:
		return value, true
	case json.Number:
		result, err := value.Float64()
		return result, err == nil
	default:
		return 0, false
	}
}

func truthy(value any) bool {
	switch value := value.(type) {
	case nil:
		return false
	case bool:
		return value
	case string:
		return value != ""
	case []any:
		return len(value) > 0
	case map[string]any:
		return len(value) > 0
	default:
		return true
	}
}

File: src/usage.rs
use serde_json::Value;

#[derive(Clone, Debug, Default, PartialEq)]
pub struct Accumulator {
    pub prompt: Option<f64>,
    pub cached: Option<f64>,
    pub completion: Option<f64>,
    pub finish_reason: String,
    pub content_chars: usize,
    pub reasoning_chars: usize,
    pub tool_call_deltas: usize,
    pub generated: bool,
}

impl Accumulator {
    pub fn feed_json(&mut self, data: &[u8]) {
        if let Ok(value) = serde_json::from_slice(data) {
            self.feed_value(&value);
        }
    }

    pub fn feed_sse_line(&mut self, line: &[u8]) {
        let line = trim_ascii(line);
        let Some(payload) = line.strip_prefix(b"data:") else {
            return;
        };
        let payload = trim_ascii(payload);
        if !payload.is_empty() && payload != b"[DONE]" {
            self.feed_json(payload);
        }
    }

    pub fn feed_value(&mut self, object: &Value) {
        if let Some(choices) = object.get("choices").and_then(Value::as_array) {
            if let Some(reason) = choices
                .first()
                .and_then(|choice| choice.get("finish_reason"))
                .and_then(Value::as_str)
            {
                reason.clone_into(&mut self.finish_reason);
            }
            for choice in choices {
                let node = choice
                    .get("delta")
                    .filter(|value| value.as_object().is_some_and(|map| !map.is_empty()))
                    .or_else(|| choice.get("message"));
                let Some(node) = node else { continue };
                if let Some(content) = node.get("content").and_then(Value::as_str) {
                    self.content_chars += content.chars().count();
                    self.generated |= !content.is_empty();
                }
                for key in ["reasoning_content", "reasoning"] {
                    if let Some(reasoning) = node.get(key).and_then(Value::as_str) {
                        self.reasoning_chars += reasoning.chars().count();
                        self.generated |= !reasoning.is_empty();
                    }
                }
                if node.get("tool_calls").is_some_and(truthy) {
                    self.tool_call_deltas += 1;
                    self.generated = true;
                }
            }
        }
        if let Some(reason) = object.get("stop_reason").filter(|value| !value.is_null()) {
            self.finish_reason = reason
                .as_str()
                .map_or_else(|| reason.to_string(), ToOwned::to_owned);
        }
        if let Some(status @ ("completed" | "incomplete" | "cancelled")) =
            object.get("status").and_then(Value::as_str)
        {
            status.clone_into(&mut self.finish_reason);
        }
        match object.get("delta") {
            Some(Value::Object(delta)) => {
                self.generated |= ["text", "thinking", "reasoning", "partial_json"]
                    .iter()
                    .any(|key| {
                        delta
                            .get(*key)
                            .and_then(Value::as_str)
                            .is_some_and(|value| !value.is_empty())
                    });
            }
            Some(Value::String(delta)) => self.generated |= !delta.is_empty(),
            _ => {}
        }

        let usage = object
            .get("message")
            .and_then(|message| message.get("usage"))
            .filter(|usage| usage.as_object().is_some_and(|map| !map.is_empty()))
            .or_else(|| object.get("usage"));
        let Some(usage) = usage else { return };
        if let Some(value) = number(usage.get("prompt_tokens")) {
            self.prompt = Some(value);
            self.cached = usage
                .get("prompt_tokens_details")
                .and_then(|details| number(details.get("cached_tokens")))
                .or_else(|| number(usage.get("cached_tokens")));
        }
        if let Some(value) = number(usage.get("completion_tokens")) {
            self.completion = Some(value);
        }
        if let Some(value) = number(usage.get("input_tokens")) {
            self.prompt = Some(value);
        }
        if let Some(value) = number(usage.get("cache_read_input_tokens")) {
            self.cached = Some(value);
        }
        if let Some(value) = number(usage.get("output_tokens")) {
            self.completion = Some(value);
        }
    }
}

pub fn feed_sse_chunk(accumulator: &mut Accumulator, tail: &mut Vec<u8>, chunk: &[u8]) {
    tail.extend_from_slice(chunk);
    let mut consumed = 0;
    while let Some(relative) = tail[consumed..].iter().position(|byte| *byte == b'\n') {
        let end = consumed + relative;
        accumulator.feed_sse_line(&tail[consumed..end]);
        consumed = end + 1;
    }
    if consumed > 0 {
        tail.drain(..consumed);
    }
}

fn trim_ascii(mut value: &[u8]) -> &[u8] {
    while value.first().is_some_and(u8::is_ascii_whitespace) {
        value = &value[1..];
    }
    while value.last().is_some_and(u8::is_ascii_whitespace) {
        value = &value[..value.len() - 1];
    }
    value
}

fn number(value: Option<&Value>) -> Option<f64> {
    value.and_then(Value::as_f64)
}

fn truthy(value: &Value) -> bool {
    match value {
        Value::Null => false,
        Value::Bool(value) => *value,
        Value::String(value) => !value.is_empty(),
        Value::Array(value) => !value.is_empty(),
        Value::Object(value) => !value.is_empty(),
        Value::Number(_) => true,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn extracts_openai_usage_and_true_output() {
        let mut usage = Accumulator::default();
        usage.feed_json(br#"{"choices":[{"message":{"content":"hi","reasoning_content":"thinking"},"finish_reason":"stop"}],"usage":{"prompt_tokens":100,"completion_tokens":5,"prompt_tokens_details":{"cached_tokens":90}}}"#);
        assert_eq!(usage.prompt, Some(100.0));
        assert_eq!(usage.cached, Some(90.0));
        assert_eq!(usage.completion, Some(5.0));
        assert_eq!(usage.content_chars, 2);
        assert_eq!(usage.reasoning_chars, 8);
        assert!(usage.generated);
    }

    #[test]
    fn reassembles_split_sse_lines() {
        let mut usage = Accumulator::default();
        let mut tail = Vec::new();
        feed_sse_chunk(
            &mut usage,
            &mut tail,
            b"data: {\"choices\":[{\"delta\":{\"content\":\"x\"}}]}\ndata: {\"usage\":{\"prompt_t",
        );
        feed_sse_chunk(
            &mut usage,
            &mut tail,
            b"okens\":10,\"completion_tokens\":3}}\n",
        );
        assert!(tail.is_empty());
        assert_eq!(usage.prompt, Some(10.0));
        assert_eq!(usage.completion, Some(3.0));
    }

    #[test]
    fn ignores_role_only_but_detects_tool_delta() {
        let mut usage = Accumulator::default();
        usage.feed_json(br#"{"choices":[{"delta":{"role":"assistant","content":""}}]}"#);
        assert!(!usage.generated);
        usage.feed_json(br#"{"choices":[{"delta":{"tool_calls":[{"index":0}]}}]}"#);
        assert!(usage.generated);
    }
}
